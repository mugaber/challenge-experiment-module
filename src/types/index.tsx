export type IterationState = 'pending' | 'done'

export type Iteration = {
  id: number
  title: string
  experimentId: number
  state: IterationState
}

export const ExperimentStatuses = {
  EMPTY: 'empty',
  LOCKED: 'locked',
  UNLOCKED: 'unlocked'
} as const

export type ExperimentStatus =
  (typeof ExperimentStatuses)[keyof typeof ExperimentStatuses]

export type Experiment = {
  id: number
  title: string
  status: ExperimentStatus
  iterations: Iteration[]
}
