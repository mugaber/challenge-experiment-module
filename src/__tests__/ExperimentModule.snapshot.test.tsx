import { describe, it, expect, vi } from 'vitest'
import ExperimentModule from '../components/experiment-module/index'
import {
  renderWithContext,
  testExperiments,
  createTestExperiment,
  createTestIteration
} from '../test/utils'
import { ExperimentStatuses, type Iteration } from '../types'

vi.mock('../assets/lock-open.svg', () => ({
  default: 'lock-open-icon'
}))
vi.mock('../assets/lock-closed.svg', () => ({
  default: 'lock-closed-icon'
}))

vi.mock('../components/Iteration', () => ({
  default: ({
    iteration,
    onlyOne,
    first,
    last
  }: {
    iteration: Iteration
    onlyOne: boolean
    first: boolean
    last: boolean
  }) => (
    <div
      data-testid={`iteration-${iteration.id}`}
      data-iteration-props={JSON.stringify({ onlyOne, first, last })}
    >
      <span>{iteration.title}</span>
      <span data-state={iteration.state}>{iteration.state}</span>
      <span data-length={iteration.length}>{iteration.length}</span>
    </div>
  )
}))

vi.mock('../components/buttons/transparent', () => ({
  default: ({
    children,
    onClick
  }: {
    children: React.ReactNode
    onClick: () => void
  }) => (
    <button onClick={onClick} data-component="transparent-button">
      {children}
    </button>
  )
}))

describe('ExperimentModule Snapshot Tests', () => {
  describe('Different States', () => {
    it('matches snapshot for empty experiment (closed)', () => {
      const { container } = renderWithContext(
        <ExperimentModule experiment={testExperiments.empty} />
      )
      expect(container.firstChild).toMatchSnapshot('empty-experiment-closed')
    })

    it('matches snapshot for empty experiment (open)', () => {
      const { container } = renderWithContext(
        <ExperimentModule experiment={testExperiments.empty} />
      )
      expect(container.firstChild).toMatchSnapshot('empty-experiment-open')
    })

    it('matches snapshot for unlocked experiment with iterations', () => {
      const { container } = renderWithContext(
        <ExperimentModule experiment={testExperiments.unlocked} />
      )
      expect(container.firstChild).toMatchSnapshot(
        'unlocked-experiment-with-iterations'
      )
    })

    it('matches snapshot for locked experiment', () => {
      const { container } = renderWithContext(
        <ExperimentModule experiment={testExperiments.locked} />
      )
      expect(container.firstChild).toMatchSnapshot('locked-experiment')
    })

    it('matches snapshot for single iteration experiment', () => {
      const singleIterationExperiment = createTestExperiment(
        1,
        ExperimentStatuses.UNLOCKED,
        'Single Iteration',
        [createTestIteration(1, 1, 'Only Iteration')]
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={singleIterationExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot(
        'single-iteration-experiment'
      )
    })
  })

  describe('Different Iteration Configurations', () => {
    it('matches snapshot for experiment with mixed iteration states', () => {
      const mixedStateExperiment = createTestExperiment(
        1,
        ExperimentStatuses.UNLOCKED,
        'Mixed States',
        [
          createTestIteration(1, 1, 'Done Iteration', 'done', 'short'),
          createTestIteration(2, 1, 'Pending Iteration', 'pending', 'medium'),
          createTestIteration(3, 1, 'Another Done', 'done', 'long')
        ]
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={mixedStateExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot('mixed-iteration-states')
    })

    it('matches snapshot for experiment with different iteration lengths', () => {
      const differentLengthsExperiment = createTestExperiment(
        1,
        ExperimentStatuses.UNLOCKED,
        'Different Lengths',
        [
          createTestIteration(1, 1, 'Short Iteration', 'done', 'short'),
          createTestIteration(2, 1, 'Medium Iteration', 'done', 'medium'),
          createTestIteration(3, 1, 'Long Iteration', 'done', 'long')
        ]
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={differentLengthsExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot(
        'different-iteration-lengths'
      )
    })
  })

  describe('Interactive States', () => {
    it('matches snapshot when adding iteration', () => {
      const { container } = renderWithContext(
        <ExperimentModule experiment={testExperiments.empty} />,
        {
          contextValue: {
            activeIteration: { iterationId: 1, experimentId: 1 }
          }
        }
      )
      expect(container.firstChild).toMatchSnapshot('adding-iteration-state')
    })

    it('matches snapshot for experiment with long title', () => {
      const longTitleExperiment = createTestExperiment(
        1,
        ExperimentStatuses.UNLOCKED,
        'Very Long Experiment Title That Might Wrap To Multiple Lines And Test Layout',
        [createTestIteration(1, 1)]
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={longTitleExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot('long-title-experiment')
    })
  })

  describe('Edge Cases', () => {
    it('matches snapshot for experiment with no title', () => {
      const noTitleExperiment = createTestExperiment(
        1,
        ExperimentStatuses.EMPTY,
        '',
        []
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={noTitleExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot('no-title-experiment')
    })

    it('matches snapshot for experiment with many iterations', () => {
      const manyIterations = Array.from({ length: 5 }, (_, i) =>
        createTestIteration(i + 1, 1, `Iteration ${i + 1}`, 'done', 'short')
      )

      const manyIterationsExperiment = createTestExperiment(
        1,
        ExperimentStatuses.UNLOCKED,
        'Many Iterations',
        manyIterations
      )

      const { container } = renderWithContext(
        <ExperimentModule experiment={manyIterationsExperiment} />
      )
      expect(container.firstChild).toMatchSnapshot('many-iterations-experiment')
    })
  })
})
