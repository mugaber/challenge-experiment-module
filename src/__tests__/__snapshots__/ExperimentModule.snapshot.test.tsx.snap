// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html


exports[`ExperimentModule Snapshot Tests > Different Iteration Configurations > matches snapshot for experiment with different iteration lengths > different-iteration-lengths 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Different Lengths
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different Iteration Configurations > matches snapshot for experiment with mixed iteration states > mixed-iteration-states 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Mixed States
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different States > matches snapshot for empty experiment (closed) > empty-experiment-closed 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white/50 cursor-pointer"
    >
      Empty Experiment
    </h3>
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different States > matches snapshot for empty experiment (open) > empty-experiment-open 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white/50 cursor-pointer"
    >
      Empty Experiment
    </h3>
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different States > matches snapshot for locked experiment > locked-experiment 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white/50 cursor-default"
    >
      Locked Experiment
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-closed-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different States > matches snapshot for single iteration experiment > single-iteration-experiment 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Single Iteration
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Different States > matches snapshot for unlocked experiment with iterations > unlocked-experiment-with-iterations 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Unlocked Experiment
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Edge Cases > matches snapshot for experiment with many iterations > many-iterations-experiment 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Many Iterations
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Edge Cases > matches snapshot for experiment with no title > no-title-experiment 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white/50 cursor-pointer"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Interactive States > matches snapshot for experiment with long title > long-title-experiment 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white cursor-pointer"
    >
      Very Long Experiment Title That Might Wrap To Multiple Lines And Test Layout
    </h3>
    <img
      alt="Lock Icon"
      class="w-6 h-6 mr-4 cursor-pointer"
      src="lock-open-icon"
    />
  </div>
</div>
`;

exports[`ExperimentModule Snapshot Tests > Interactive States > matches snapshot when adding iteration > adding-iteration-state 1`] = `
<div
  class="flex flex-col items-center justify-center px-4 py-6 gap-4
        bg-[#242424] rounded-lg w-11/12 sm:w-8/12 lg:w-6/12 2xl:w-4/12"
>
  <div
    class="w-full flex justify-between items-center"
  >
    <h3
      class="flex-1 text-2xl font-bold text-white/50 cursor-pointer"
    >
      Empty Experiment
    </h3>
  </div>
</div>
`;
